### PR TITLE
get data by ssl authentication

### PR DESCRIPTION
**Pull request by pronix**
using theforeman as data source(without puppetdb). using ssl keys of foreman proxy for authentication.
require also change in foreman file ./app/controllers/api/v2/hosts_controller.rb
```
add_smart_proxy_filters :facts, :features => Proc.new { FactImporter.fact_features }
```
to
```
add_smart_proxy_filters [:facts,:index], :features => Proc.new { FactImporter.fact_features }
```

### CODE REVIEW

File: lib/puppet/parser/functions/foreman.rb [hunk 1]
```diff
@@ -64,14 +66,24 @@ module Puppet::Parser::Functions
...
 
-      http = Net::HTTP.new(uri.host, uri.port)
+      if use_tfmproxy
+        tfmproxy = YAML.load(File.read('/etc/foreman-proxy/settings.yml'))
+        uri = URI.parse(tfmproxy[:foreman_url])
+        http = Net::HTTP.new(uri.host, uri.port)
+        http.ca_file = tfmproxy[:foreman_ssl_ca]
+        http.cert = OpenSSL::X509::Certificate.new(File.read(tfmproxy[:foreman_ssl_cert]))
+        http.key = OpenSSL::PKey::RSA.new(File.read(tfmproxy[:foreman_ssl_key]), nil)
+        http.verify_mode = OpenSSL::SSL::VERIFY_NONE
```
Comment (ekohl) [inline lib/puppet/parser/functions/foreman.rb]: Why do you configure a CA but don't verify?

**Comment by pronix** [comment]
hello @ekohl 
thanks for review.
peer verification is changed and use_tfmproxy can be a string for custom file.
but i still don't want to change current behaviour.

File: lib/puppet/parser/functions/foreman.rb [hunk 2]
```diff
@@ -64,14 +66,29 @@ module Puppet::Parser::Functions
...
-      uri = URI.parse(foreman_url)
 
       req = Net::HTTP::Get.new(path)
-      req.basic_auth(foreman_user, foreman_pass)
       req['Content-Type'] = 'application/json'
       req['Accept'] = 'application/json'
 
-      http = Net::HTTP.new(uri.host, uri.port)
+      if use_tfmproxy
+        configfile = if use_tfmproxy.is_a? String and File.exist?(use_tfmproxy)
```
Comment (ekohl) [inline lib/puppet/parser/functions/foreman.rb]: I'd move the ```File.exist?``` to just before reading it. If a user specifies a non-existing file it shouldn't silently ignore it. The same is true for ```true``` but ```/etc/foreman-proxy/settings.yml``` doesn't exist.

Comment (pronix) [reply lib/puppet/parser/functions/foreman.rb]: you mean exception or just puts ?

Comment (ekohl) [reply lib/puppet/parser/functions/foreman.rb]: I think a hard exception is in order.

**Comment by pronix** [comment]
hello @ekohl 
added files checking before read.

File: lib/puppet/parser/functions/foreman.rb [hunk 3]
```diff
@@ -64,14 +66,36 @@ module Puppet::Parser::Functions
...
 
       req = Net::HTTP::Get.new(path)
-      req.basic_auth(foreman_user, foreman_pass)
       req['Content-Type'] = 'application/json'
       req['Accept'] = 'application/json'
 
-      http = Net::HTTP.new(uri.host, uri.port)
+      if use_tfmproxy
+        if use_tfmproxy.is_a? String
+          if File.exist?(use_tfmproxy)
```
Comment (ekohl) [inline lib/puppet/parser/functions/foreman.rb]: This block can be simplified I think:
```ruby
configfile = '/etc/foreman-proxy/settings.yaml'
configfile = use_tfmproxy if use_tfmproxy.is_a? String
raise Puppet::ParseError, "File #{configfile} not found while use_tfmproxy is enabled" unless File.exists?(configfile)
```

**Comment by pronix** [comment]
thanks @ekohl . you are absolutely right.

File: lib/puppet/parser/functions/foreman.rb [hunk 4]
```diff
@@ -64,14 +66,27 @@ module Puppet::Parser::Functions
...
+        http.cert = OpenSSL::X509::Certificate.new(File.read(tfmproxy[:foreman_ssl_cert]))
+        http.key = OpenSSL::PKey::RSA.new(File.read(tfmproxy[:foreman_ssl_key]), nil)
+        http.verify_mode = OpenSSL::SSL::VERIFY_PEER
+      else
+        uri = URI.parse(foreman_url)
+        http = Net::HTTP.new(uri.host, uri.port)
+        req.basic_auth(foreman_user, foreman_pass)
+      end
       http.use_ssl = true if uri.scheme == 'https'
       http.verify_mode = OpenSSL::SSL::VERIFY_NONE if http.use_ssl?
```
Comment (ekohl) [inline lib/puppet/parser/functions/foreman.rb]: I think this should move to the above ```else``` section.

**Comment by pronix** [comment]
thanks @ekohl . fixed

**Comment by mmoll** [comment]
I can't really this this, but it looks like if all fallback mechanisms are there, so merging...

**Comment by mmoll** [comment]
merged, спасибо @pronix!

**Pull request merged**